ddtrace/internal/telemetry/telemetry_writer.py

```diff
@@ -0,0 +1,239 @@
+    MAX_FAIL_COUNT = 3
+
+    enabled = False  # type: ClassVar[bool]
+    _instance = None  # type: ClassVar[Optional[TelemetryWriter]]
+    _lock = forksafe.Lock()  # type: ClassVar[forksafe.ResetObject]
+    sequence = 0  # type: ClassVar[int]
+
+    def __init__(self):
+        # type: () -> None
+        super(TelemetryWriter, self).__init__(interval=_get_interval_or_default())
+
+        self.encoder = JSONEncoderV2()
+        self._events_queue = []  # type: List[TelemetryRequest]
+        self._integrations_queue = []  # type: List[Dict]
+
+    def _get_headers(self, payload_type):
+        # type: (str) -> Dict
+        """
+        Creates request headers
+        """
+        return {
+            "Content-type": "application/json",
+            "DD-Telemetry-Request-Type": payload_type,
+            "DD-Telemetry-API-Version": "v2",
+        }
+
+    def _send_request(self, request):
+        # type: (Dict) -> httplib.HTTPResponse
+        """
+        Sends a telemetry request to the trace agent
+        """
+        conn = get_connection(self.AGENT_URL)
+        rb_json = self.encoder.encode(request)
+        resp = None
+        with StopWatch() as sw:
+            try:
+                conn.request("POST", self.ENDPOINT, rb_json, self._get_headers(request["request_type"]))
+                resp = get_connection_response(conn)
+
+                t = sw.elapsed()
+                if t >= self.interval:
+                    log_level = logging.WARNING
+                else:
+                    log_level = logging.DEBUG
+
+                log.log(
+                    log_level,
+                    "sent %d in %.5fs to %s/%s. response: %s",
+                    len(rb_json),
+                    t,
+                    self.AGENT_URL,
+                    self.ENDPOINT,
+                    resp.status,
+                )
+            finally:
+                conn.close()
+        return resp
+
+    def flush_integrations_queue(self):
+        # type () -> List[Dict]
+        """Returns a list of all integrations queued by add_integration"""
+        with self._lock:
+            integrations = self._integrations_queue
+            self._integrations_queue = []
+        return integrations
+
+    def flush_events_queue(self):
+        # type () -> List[TelemetryRequest]
+        """Returns a list of all integrations queued by classmethods"""
+        with self._lock:
+            requests = self._events_queue
+            self._events_queue = []
+        return requests
+
+    def periodic(self):
+        integrations = self.flush_integrations_queue()
+        if integrations:
+            self.app_integrations_changed_event(integrations)
+
+        telemetry_requests = self.flush_events_queue()
+
+        requests_failed = []  # type: List[TelemetryRequest]
+        for telemetry_request in telemetry_requests:
+            request, fail_count = telemetry_request
+            resp = self._send_request(request)
```

Comment:
No exceptions are handled around the http sending which can raise these exceptions: https://github.com/DataDog/dd-trace-py/blob/a25533ea93cdf98bcd1916e3a26e2c4650778b9b/ddtrace/internal/writer.py#L302

Not sure if we want to retry the requests or not on failure.

Reply:
I'll catch all exceptions and just drop the failing payload